Dictionary Service to generate HTML via Spring MVC, Data Dictionary page Enhancements

kevinrr888 opened the pull request:
Dictionary Service should no longer use HtmlProvider interface and a stringbuilder to build the HTML. Should use Spring MVC instead. ([Datawave Issue 1598](https://github.com/NationalSecurityAgency/datawave/issues/1598)). This PR also fixes [Datawave Issue 1634](https://github.com/NationalSecurityAgency/datawave/issues/1634).

Changes:
- HTML for the Data Dictionary, Edge Dictionary, Field Description, and Model pages are no longer served by HtmlProvider.
  Instead, Spring MVC with HTML template files are used.
- DataDictionaryController, EdgeDictionaryController, and ModelController no longer handle HTML. Handled by
  new classes DataDictionaryWebController, EdgeDictionaryWebController, and ModelWebController now.
- Added thymeleaf dependency to service/pom.xml
- Data Dictionary page columns can now be resized by the user and page now shows how the table is being sorted.

Model Pages (/list and /{name}) specifically:

- Fixed issue with jQuery scripts not being loaded properly on these pages (/list and /{name})
- Entries of table in /list were originally linking to "/DataWave/Model/{name}?modelTableName={modelTableName}" which I believe doesn't exist. I think this was meant to link to the /{name} endpoint (the Model Description page) so I changed it to "/model/{name}?name={name}&modelTableName={modelTableName}". This may not have been what was intended so let me know.

kevinrr888 committed:
Data Dictionary page to no longer use HtmlProvider

- HTML for the Data Dictionary page is no longer served by HtmlProvider.
  Instead, Spring MVC with a template HTML file is used.
- Added thymeleaf dependency to service/pom.xml
- DataDictionaryController no longer handles HTML. Handled by
  new class DataDictionaryWebController now.

kevinrr888 committed:
Simplified datadictionary.html header scripts

Figured out simpler way to dynamically source the script tags by simply
putting "th:src='${LINK_HERE}'". No longer need to wait on load events.

kevinrr888 committed:
Edge Dictionary Page Now Uses Spring MVC:

- Changed Edge Dictionary page to no longer rely on HtmlProvider: now
  uses Spring MVC with template file.
- Added EdgeDictionaryWebController to handle the html requests for the Edge Dictionary page instead of EdgeDictionaryController.
- Fixed formatting issue for the Data Dictionary page (wasn't exactly matching orignal page after moving to template file). For some reason, <!DOCTYPE html> was the cause.

kevinrr888 committed:
Field Description pages now use Spring MVC:

- The Field Description pages (/Descriptions/,
  /Descriptions/{datatype}/, and /Descriptions/{datatype}/{fieldName}/)
  now use Spring MVC instead of HtmlProvider.

jwomeara commented on service/src/main/java/datawave/microservice/dictionary/DataDictionaryWebController.java:
```diff
@@ -0,0 +1,275 @@
+        
+        ModelAndView mav = new ModelAndView();
+        mav.setViewName("datadictionary");
+        mav.addObject("jqueryUri", dataDictionary.getJqueryUri() + "jquery.min.js");
+        mav.addObject("datatablesUri", dataDictionary.getDatatablesUri() + "jquery.dataTables.min.js");
```
I know the idea of passing in the jquery and datatables URI predates your work, but we could avoid passing these URIs around in java by using a thymeleaf placeholder in the html where they are used.

These properties are ultimately coming from our spring configuration/properties for the dictionary service.  These properties can be accessed directly in the html via thymeleaf.  Here's an example where I did this for the query service:

https://github.com/NationalSecurityAgency/datawave-query-service/blob/67ca1e103db7ab39450a35cea3fac4417a163d9b/service/src/main/resources/templates/index.html#L27

These values are constructed using 'server.cdnUri' from the DatawaveServerProperties in the datawave spring boot starter.  

If you were to access that property directly via thymeleaf, it would look something like this in your HTML:
${@environment.getProperty('server.cdnUri')webjars/jquery/jquery.min.js}
${@environment.getProperty('server.cdnUri')webjars/datawables/js/jquery.dataTables.min.js}

I wouldn't be opposed to having a dictionary-specific override value for the webjars locations though in the event that we don't want to use the cdnUri property for some reason.  Thoughts?

kevinrr888 replied:
Ahh very nice. Thanks! I think it would probably be cleaner to do it your way with accessing it directly in thymeleaf so we aren't unnecessarily passing around these strings (also removes the need for a couple getters I made). Let me know if I should change this or keep what I have now.

jwomeara commented on service/src/main/java/datawave/microservice/dictionary/DataDictionaryWebController.java:
```diff
@@ -0,0 +1,275 @@
+        
+        // To be passed to the MAV. Contains all the table content
+        List<List<String>> tableContent = new ArrayList<List<String>>();
+        for (META f : fields) {
```
Just curious whether or not you ever tried adding the fields variable directly to the model, and whether that worked.  If that worked, would you be able to avoid some of this transformation logic?

kevinrr888 replied:
Yes that is what I initially tried but passing it as is to the model, it wasn't an iterable array (had some funky, unnecessary data)... I assume this was probably the result of the toString() methods for the data.

jwomeara commented on service/src/main/java/datawave/microservice/dictionary/DataDictionaryWebController.java:
```diff
@@ -0,0 +1,275 @@
+     */
+    @GetMapping("/")
+    @Timed(name = "dw.dictionary.data.get", absolute = true)
+    public ModelAndView get(@RequestParam(required = false) String modelName, @RequestParam(required = false) String modelTableName,
+                    @RequestParam(required = false) String metadataTableName, @RequestParam(name = "auths", required = false) String queryAuthorizations,
+                    @RequestParam(defaultValue = "") String dataTypeFilters, @AuthenticationPrincipal ProxiedUserDetails currentUser) throws Exception {
```
FYI, another way to do MVC is to just return a string from the rest endpoint which represents the view name, and add a `Model` argument to the rest endpoint signature, which is where you would set all of your model fields.  Not asking you to change anything - just sharing info.

kevinrr888 replied:
Thank you for the info! Yeah, I read about doing it like that but I was afraid of changing the method signature haha

jwomeara commented on service/src/main/resources/templates/datadictionary.html:
```diff
@@ -0,0 +1,70 @@
+				"bPaginate": false, 
+				"bStateSave": true
+			});
+			setTimeout( function() { 
+				$('#myTable').find("td").css("word-break", "break-word"); 
+			}, 4000); 
```
Ha! You removed this in the other PR, right?

kevinrr888 replied:
Haha yes! I planned on changing it for this PR too but guess I forgot to go back and delete it. Seems to be still working as expected when I delete it so I'm just going to remove it. I wonder why there was this 4 second wait time to begin with :thinking:

kevinrr888 committed:
Added Requested Changes:

- Merged in https://github.com/NationalSecurityAgency/datawave-dictionary-service/pull/15/ but instead used jQueries "resizable"
  function to make the columns resizable (simpler, works better that my original attempt).
- No longer pass the URIs for jQuery and datatables around. Access directly now via environment.getProperty('server.cdnUri')
  in thymeleaf (this removed the need for a couple getters and constructors in DefaultDataDictionary).

kevinrr888 committed:
Fixed "aaSorting" issue

For some reason, 2d array literals in thymeleaf must be formatted on
multiple lines otherwise you get an internal server error

kevinrr888 committed:
Model pages now use Spring MVC

- Model pages (/list and /{name}) no longer use HtmlProvider. Instead use Spring MVC.
- Fixed issue with jQuery scripts not being loaded properly on these pages (/list and /{name})
- Entries of table in /list were originally linking to /DataWave/Model/<name>?modelTableName=<modelTableName> which does not exist. I think this was meant to link to the /{name} endpoint (the Model Description page) so I changed it to /model/<name>?name=<name>&modelTableName=<modelTableName>.

kevinrr888 commented:
Finished replacing HtmlProvider with Spring MVC.
List of changes in PR:
- HTML for the Data Dictionary, Edge Dictionary, Field Description, and Model pages are no longer served by HtmlProvider.
Instead, Spring MVC with HTML template files are used.
- DataDictionaryController, EdgeDictionaryController, and ModelController no longer handle HTML. Handled by
new classes DataDictionaryWebController, EdgeDictionaryWebController, and ModelWebController now.
- Added thymeleaf dependency to service/pom.xml
- Data Dictionary page columns can now be resized by the user and page now shows how the table is being sorted.

Model Pages (/list and /{name}) specifically:
- Fixed issue with jQuery scripts not being loaded properly on these pages (/list and /{name})
- Entries of table in /list were originally linking to "/DataWave/Model/{name}?modelTableName={modelTableName}" which I believe doesn't exist. I think this was meant to link to the /{name} endpoint (the Model Description page) so I changed it to "/model/{name}?name={name}&modelTableName={modelTableName}". This may not have been what was intended so let me know.
